apps/system/js/audio_channel_manager.js

```diff
@@ -0,0 +1,267 @@
+/* global BaseModule */
+'use strict';
+
+(function() {
+  var PLAY_CURRENT_AND_NEW_CHANNEL = 0;
+  var PAUSE_CURRENT_CHANNEL = 1;
+  var PAUSE_THEN_RESUME_CURRENT_CHANNEL = 2;
+  var VOLUME_DOWN_CURRENT_CHANNEL = 3;
+  var VOLUME_DOWN_NEW_CHANNEL = 4;
+  var PAUSE_THEN_VIBRATE_CURRENT_CHANNEL = 5;
+  var VIBRATE_NEW_CHANNEL = 6;
+
+  var AudioChannelManager = function() {};
+
+  BaseModule.create(AudioChannelManager, {
+    name: 'AudioChannelManager',
+    DEBUG: false,
+    _currentChannels: null,
+    _pausedChannels: null,
+    _smallerVolumeChannels: null,
+
+    /**
+     * Initial the module.
+     */
+    _start: function() {
+      this._currentChannels = [];
+      this._pausedChannels = [];
+      this._smallerVolumeChannels = [];
+    },
+
+    /**
+     * Add an app into manager to manage its audio channels.
+     * @param {MozBrowser} app An app you want to manage.
+     */
+    register: function(app) {
+      app.allowedAudioChannels.forEach(function(channel) {
+        channel.onactivestatechanged = function(evt) {
+          var channel = evt.target;
+          if (channel.isActive()) {
+            this._handleAudioChannelCompeting(channel);
+          } else {
+            this._resumeLatestPausedChannel();
+            this._setVolume(this._smallerVolumeChannels, 1);
+          }
+        };
+      });
+    },
+
+    /**
+     * Remove an app from manager,
+     * if you don't wnat to manage it anymore.
+     * @param {MozBrowser} app An app you want to remove.
+     */
+    unregister: function(app) {
+      app.allowedAudioChannels.forEach(function(channel) {
+        channel.onactivestatechanged = undefined;
+      });
+    },
+
+    /**
+     * Handel the new audio channel.
+     * @param {BrowserElementAudioChannel} channel 
+     * A New channel you want to handel.
+     */
+    _handleAudioChannelCompeting: function(channel) {
+      var currentChannel;
+      switch (this.compete(channel.name)) {
+        case PLAY_CURRENT_AND_NEW_CHANNEL:
+          this._play(channel);
+          this._currentChannels.push(channel);
+          break;
+
+        case PAUSE_CURRENT_CHANNEL:
+          currentChannel = this._currentChannels.pop();
+          this._pause(currentChannel);
+          break;
+
+        case PAUSE_THEN_RESUME_CURRENT_CHANNEL:
+          currentChannel = this._currentChannels.pop();
+          this._pause(currentChannel);
+          this._pausedApps.push(currentChannel);
+          break;
+
+        case VOLUME_DOWN_CURRENT_CHANNEL:
```

Comment:
confusing naming; it's faded but the method is volume down. FADE_OUT_CURRENT_CHANNEL maybe.